lib/changereport.go

```diff
@@ -0,0 +1,460 @@
+		// 		deltaCol["median"] = -res.Columns[i].Left.(map[string]interface{})["median"].(float64)
+		// 		deltaCol["mean"] = -res.Columns[i].Left.(map[string]interface{})["mean"].(float64)
+		// 	} else if leftColItems[v.LPos].Type == "string" {
+
+		// 	} else if leftColItems[v.LPos].Type == "bool" {
+
+		// 	}
+		// } else if v.RPos >= 0 {
+		// 	if rightColItems[v.RPos].Type == "number" || rightColItems[v.RPos].Type == "integer" {
+		// 		deltaCol["count"] = res.Columns[i].Right.(map[string]interface{})["count"].(int)
+		// 		deltaCol["max"] = res.Columns[i].Right.(map[string]interface{})["max"].(float64)
+		// 		deltaCol["min"] = res.Columns[i].Right.(map[string]interface{})["min"].(float64)
+		// 		deltaCol["median"] = res.Columns[i].Right.(map[string]interface{})["median"].(float64)
+		// 		deltaCol["mean"] = res.Columns[i].Right.(map[string]interface{})["mean"].(float64)
+		// 	} else if leftColItems[v.RPos].Type == "string" {
+
+		// 	} else if leftColItems[v.RPos].Type == "bool" {
+
+		// 	}
+		// }
+
+		res.Columns[i].Delta = deltaCol
+		i++
+	}
+
+	return res, nil
+}
+
+// This takes some assumptions - we work only with tabular data, with header rows and functional structure.json
+// ChangeReport computes the change report of two sources
+func (m *DatasetMethods) ChangeReport(p *ChangeReportParams, res *ChangeReportResponse) (err error) {
+	// absolutize any local paths before a possible trip over RPC to another local process
+	if !dsref.IsRefString(p.LeftSide) {
+		if err = qfs.AbsPath(&p.LeftSide); err != nil {
+			return err
+		}
+	}
+	if !dsref.IsRefString(p.RightSide) {
+		if err = qfs.AbsPath(&p.RightSide); err != nil {
+			return err
+		}
+	}
+
+	if m.inst.rpc != nil {
+		return checkRPCError(m.inst.rpc.Call("DatasetMethods.ChangeReport", p, res))
+	}
+
+	ctx := context.TODO()
+
+	// Left side of diff loaded into a component
+	parseResolveLoad, err := m.inst.NewParseResolveLoadFunc(p.Remote)
+	if err != nil {
+		return err
+	}
+	leftDs, err := parseResolveLoad(ctx, p.LeftSide)
+	if err != nil {
+		if errors.Is(err, dsref.ErrNoHistory) {
+			return qerr.New(err, fmt.Sprintf("dataset %s has no versions, nothing to diff against", p.LeftSide))
+		}
+		return err
+	}
+	rightDs, err := parseResolveLoad(ctx, p.RightSide)
+	if err != nil {
+		if errors.Is(err, dsref.ErrNoHistory) {
+			return qerr.New(err, fmt.Sprintf("dataset %s has no versions, nothing to diff against", p.RightSide))
+		}
+		return err
+	}
+
+	ldsStr, err := json.Marshal(leftDs)
+	if err != nil {
+		log.Errorf("failed to marshal leftDs")
+	} else {
+		log.Debugf("LoadDataset: %s", ldsStr)
+	}
+
+	rdsStr, err := json.Marshal(rightDs)
+	if err != nil {
+		log.Errorf("failed to marshal rightDs")
+	} else {
+		log.Debugf("LoadDataset: %s", rdsStr)
+	}
+
+	err = inlineScriptsToBytes(leftDs)
+	if err != nil {
+		return err
+	}
+	err = inlineScriptsToBytes(rightDs)
+	if err != nil {
+		return err
+	}
+
+	if leftDs.Commit != nil || rightDs.Commit != nil {
```

Comment:
Given that all of these fields have the same type, perhaps, it's better to use a map from "component name" -> ChangeReportComponent? There doesn't seem to be a need from the type system to unfold all of these blocks for the sake of type-safety. If the problem is that the `leftDs` and `rightDs` have strongly typed structs, perhaps `base/component/component.go` can help, it provides a type-safe way to generically map over the dataset components. It's also supposed to abstract over datasets being from the repository vs FSI, but there's some bugs in how it does so.

Reply:
Nice one, I'll use the component map and refactor this to be a bit more streamlined! The component compare is handy.